spec.bs

```diff
@@ -1856,55 +1875,94 @@ To <dfn>convert to an AuctionAd sequence</dfn> given a [=list=]-or-null |ads|:
-
-
-  The network response has to be parsed to pull out the pieces relevant to each
-  [=evaluating a scoring script|evaluation of a scoring script=].
-1. Let |trustedScoringSignals| be null.
-1. Let «|allTrustedScoringSignals|, ignored, |scoringDataVersion|» be the result of [=fetching
-  trusted signals=] with |fullSignalsUrl| and false.
-1. If |allTrustedScoringSignals| is an [=ordered map=]:
-  1. Set |trustedScoringSignals| to a new empty [=map=].
-  1. [=map/Set=] |trustedScoringSignals|["`renderURL`"] to a new empty [=map=].
-  1. If |allTrustedScoringSignals|["`renderURLs`"] [=map/exists=] and
-    |allTrustedScoringSignals|["`renderURLs`"][|renderURL|] [=map/exists=], then [=map/set=] |trustedScoringSignals|["`renderURL`"][|renderURL|] to
-    |allTrustedScoringSignals|["`renderURLs`"][|renderURL|].
-  1. If |adComponentRenderURLs| is not [=list/empty=]:
-    1. Let |adComponentRenderURLsValue| be a new empty [=map=].
-    1. If |allTrustedScoringSignals|["`adComponentRenderURLs`"] [=map/exists=], [=set/for each=]
-      |adComponentRenderURL| in |adComponentRenderURLs|:
-      1. If |allTrustedScoringSignals|["`adComponentRenderURLs`"][|adComponentRenderURL|]
-        [=map/exists=], then [=map/set=] |adComponentRenderURLsValue|[|adComponentRenderURL|] to
-        |allTrustedScoringSignals|["`adComponentRenderURLs`"][|adComponentRenderURL|].
-    1. [=map/Set=] |trustedScoringSignals|["`adComponentRenderURLs`"] to |adComponentRenderURLsValue|.
+1. Let |fullSignalsUrl| be null.
+1. If |auctionConfig|'s [=auction config/trusted scoring signals url=] is not null:
+  1. Let |fullSignalsUrl| be the result of [=building trusted scoring signals url=] with |auctionConfig|'s
+    [=auction config/trusted scoring signals url=], «|renderURL|», |adComponentRenderURLs|,
+    |auctionConfig|'s [=auction config/seller experiment group id=], and |topLevelOrigin|.
+
+    Implementations may batch requests by collecting render URLs and ad component render URLs
+    from multiple invocations of [=score and rank a bid=] and passing them all to a single invocation
+    of [=building trusted scoring signals url=] to get a |scoringSignalsUrl|. By employing this approach,
+    the [=string/length=] of [=URL serializer|serialized=] |scoringSignalsUrl| must not exceed the
+    [=auction config/max trusted scoring signals url length=] of the auction. In cases where the length
+    limit is exceeded, the request must be divided into smaller pieces to comply with the length restriction.
+    However, note that a single request is always considered valid, regardless of whether its length exceeds
+    the auction's [=auction config/max trusted scoring signals url length=].
+
+    The network response has to be parsed to pull out the pieces relevant to each
+    [=evaluating a scoring script|evaluation of a scoring script=].
+
+    These requests may also begin before the script fetch, but requests cross-origin to the

```

Comment:
It's the thing that computes the allowCrossOriginTrustedScoringSignalsFrom argument to the function. 
This is called  [=generate and score bids=] => [=score and rank a bid=] => [= fetch and decode trusted scoring signals=], first the first having multiple instances.

The corresponding fetch is from the: 
Let « decisionLogicScript, allowCrossOriginTrustedScoringSignalsFrom » be the result of [fetching script] 
... which is 3rd step of [=generate and score bids=]

...Anyway, the data flow on allowCrossOriginTrustedScoringSignalsFrom suggests that this at least isn't time travelling...

Reply:
OK I think I understand what's going on here but let me try to verify my understanding with you.

1. First, can you please clarify (maybe here but mainly in the spec) what "These requests" refers to? It is not clear. I think it is referring to the single request below, in step 10.2, "fetching trusted signals with fullSignalsUrl". Maybe making that step linkable, and then linking to it from here would be better. Or just using more specific wording.

1. There can be **many** instances of [generate and score bids][1]. Each one's _step 3_ retrieves a `« decisionLogicScript, allowCrossOriginTrustedScoringSignalsFrom »`. Then way later down the chain, it eventually invokes `[=fetch and decode trusted scoring signals=]` which "may" batch **many** calls to both `[=building trusted scoring signals url=]` and the actual fetch in 10.2 (`the result of fetching trusted signals with fullSignalsUrl, [...]`).

    But then further down I read "These requests may also begin before the script fetch". Assuming "these requests" refers to step 10.2, then how is it even possible for 10.2 run executed before ["the script fetch"][2], which happens much earlier in the callstack?

1. *Unrelated to this PR*: What does `|scoringSignalsUrl|` refer to? Is it just a stand-in for `|fullSignalsUrl|` in a world where the batching is done, and only a single call to `[=building trusted scoring signals URL=]` (and 10.2) is finally made?
    1. Separately, the whole part from "By employing this approach" and beyond seems contradictory, no? It says the final signals URL length must be under a certain length to be valid (if it's too large, must be broken up). But then it says "a single request is always valid, regardless of length". That last part makes me think there is no need to ever break the URL up into multiple requests, since "a single request is _always_ valid". What am I missing?

(2) is the most important part of this comment I think. But even with that addressed, I think this "may" is getting too complicated and unclear to continue without being fully specified. Especially with the difference in cross-origin handling / header parsing. I think we'll need to spec out the model of batching where we have some global-ish structure where batched URLs are optionally stored, combined, and finally fetched. That will make whatever relationship that exists between "the script request" (generate & score > step 3) and the signals request (10.2 of this algorithm) much more clear.

[1]: https://pr-preview.s3.amazonaws.com/morlovich/turtledove/pull/1197.html#generate-and-score-bids
[2]: https://pr-preview.s3.amazonaws.com/morlovich/turtledove/pull/1197.html#generate-and-score-bids:~:text=Let%20%C2%AB%20decisionLogicScript%2C%20allowCrossOriginTrustedScoringSignalsFrom%20%C2%BB%20be%20the%20result%20of%20fetching%20script%20with%20auctionConfig%E2%80%99s%20decision%20logic%20url

Reply:
for 3i: multiple requests can be combined, but when building the combined fetch URL from multiple request URLs, it cannot exceed the length limit (otherwise combine less requests to this one, e.g., if combining 3 requests results in a too long URL, then try to only combine 2).
The last sentence is about "any single request (before combining) is always treated as valid", because a single request cannot be broken down.

Reply:
1) Would saying "trusted scoring signals requests" help, which are indeed those that happen in 10.2?
   Though it does need to say which ones can be merged (they must have at least the same auctionConfig’s 
    trusted scoring signals url, auctionConfig’s seller experiment group id, and topLevelOrigin mechanically; 
    and  I am not sure as to all the implications for merging with e.g. different sellers.

   ... how do I make the step linkable?
 
2)  ... If allowCrossOriginTrustedScoringSignalsFrom is not needed (something that can be determined the 
     moment configuration is first parsed!), e.g. in same-origin case, all the data  required to make those 
    requests is completely independent from the script seller fetch step (but is dependent on completion 
    of bidder computation) .  Ultimately, what's happening here is that the spec writes "fetch a script" as a 
    blocking operation followed a while later by a bunch of "fetch trusted signals" operations to keep things 
    simple, while any sensible implementation  would have  those as partially overlapped asynchronous 
    operations where we need both to succeed by the time we actually run the script.

3) Yes.  
     i. A single URL is never broken up; if it's beyond the limit we will use as is, but we will not merging 
     anything into it.

Overall, I generally agree with your opinion, but I am not sure this CL is the place to do it (in part because I have no idea on how to do it in the first place). Note that any definition will be non-deterministic, and there is no reason to expect the same merging heuristic we use (a 10ms window, basically) to be used by other people.

Reply:
> Would saying "trusted scoring signals requests" help, which are indeed those that happen in 10.2?

I think that sounds better? But that plus making it linkable would make it 100% clear, so it might be better.

> ... how do I make the step linkable?

https://github.com/WICG/turtledove/blob/main/spec.bs#L3671

Regarding your (2) above:

> while any sensible implementation would have those as partially overlapped asynchronous
operations where we need both to succeed by the time we actually run the script.

This is exactly what we should spec. What a sensible implementation would do.

> Note that any definition will be non-deterministic

It doesn't have to be deterministic. The scaffolding just has to exist enough to the point where the non-determinism can be slotted in at places that make sense and can be replicated. This is in contrast with the current spec, where the whole mechanics and flow of batching and merging are just completely mysterious and probably not cleanly replicable by reading the spec.

If you absolutely have to, we can do this in a follow-up. But I think after this PR the state of things would be rough enough to where I'd probably want that follow-up to be a P0 before reviewing other new work.

Reply:
> > Would saying "trusted scoring signals requests" help, which are indeed those that happen in 10.2?
> 
> I think that sounds better? But that plus making it linkable would make it 100% clear, so it might be better.

Tried to do both and to also rephrase the URL limit stuff.

> > Note that any definition will be non-deterministic
> 
> It doesn't have to be deterministic. The scaffolding just has to exist enough to the point where the non-determinism can be slotted in at places that make sense and can be replicated. This is in contrast with the current spec, where the whole mechanics and flow of batching and merging are just completely mysterious and probably not cleanly replicable by reading the spec.
> 
> If you absolutely have to, we can do this in a follow-up. But I think after this PR the state of things would be rough enough to where I'd probably want that follow-up to be a P0 before reviewing other new work.

Well, I don't see any ways of doing this that don't involve completely reworking the entire auction execution portion of the spec.

Reply:
I don't know if it would really require reworking the _entire_ auction execution portion... but I'm not positive. I think it's very important that the spec match reality though. The point of it is to make sure we produce something that can be re-implemented by another engine, and I don't feel confident that piling onto the current request batching handwave makes that realistic.

Reply:
So how exactly does one define things that are equivalent to the specification? i.e. normally those sorts of things tend to be vaguely of the form "computation result is the same, and side effects are the same", but how do you define side effect equivalence here?

e.g. is there an ordering on fetches? (I hope not, since it's very hard to say which order they happen in an inter-process system like Chrome, and reasoning about partial orders of them seems head-blowing-up hard).

This isn't just about trusted signals, either --- e.g. the way this is written out fetches scripts before every execution thereof, for simplicity, but the implementation doesn't do that. 

What the current spec does well, I think, is to explain how to get to the result.

Reply:
Thought: so for trusted scoring signals fetches alone, they could be modeled as some sort of a cross-thread queue were requests are added, which non-deterministically grabs them and merges then sends back replies. I have no clue on how to build it in spec'ese, and that still requires some sort of understanding about what sorts of re-orderings are permissible..

Reply:
Can you make those PRs go against the actual live repository, instead of the fork? If we merge those forked PRs, I don't think they'll end up here.

Reply:
Then I assume we can land those and revisit this? Also documenting each PR in the description is going to go a long way.

Reply:
If this lands I can make them against the main repository. Well, or rather make a different pull from the same branch in my work. If I make a pull request against it now it will include this change and be completely unreadable.

> Then I assume we can land those and revisit this? Also documenting each PR in the description is going to go a long way.

Those are based on this CL, not the other way around. (And the second entirely replaces the batching handling).

Reply:
OK that makes sense, thanks a lot of getting those two PRs going so quickly, I think it'll go a long way towards seriously hardening this stuff up. I think we're probably good to proceed with this PR now.